### PR TITLE
move to GitHub Actions

### PR DESCRIPTION
**Pull request by kmoe**

### CODE REVIEW

**Comment by kmoe** [comment]
looking into test failure

**Review by paultyng — APPROVED** [review]

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+
+    - name: Set up Go
+      uses: actions/setup-go@v2.1.3
+      with:
+        go-version: '1.15'
```
Comment (bflad) [inline .github/workflows/test.yml:25]:
```suggestion
        go-version: '1.16'
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 15
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [macos-latest, windows-latest, ubuntu-latest]
+        terraform:
+          - '0.12.30'
+          - '0.13.6'
+          - '0.14.5'
```
Comment (bflad) [inline .github/workflows/test.yml:57]:
```suggestion
          - '0.12.31'
          - '0.13.7'
          - '0.14.11'
          - '0.15.1'
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    steps:
+
+    - name: Set up Go
+      uses: actions/setup-go@v2.1.3
+      with:
+        go-version: '1.15'
```
Comment (bflad) [inline .github/workflows/test.yml:63]:
```suggestion
        go-version: '1.16'
```

**Review by bflad — COMMENTED** [review]
Since this was raised, there are some newer versions of Go and Terraform available 🎉

**Comment by bflad** [comment]
Now we are getting to the good stuff!

Linux and macOS - Terraform CLI 0.13.7:

Reference: https://github.com/hashicorp/terraform/issues/25961

Essentially, Terraform CLI 0.13 is different than other versions of Terraform CLI w.r.t. data source ordering in the graph. This would require any one of these fixes:

- Creating a `testdata/` file in the repository and referencing that path in the configuration instead of using the `local_file` resource
- Splitting the test steps into two steps (first to apply just the resource and second to apply both)

Adding the required `depends_on = [local_file.file]` in the data source for Terraform CLI 0.13 I believe would break other versions, so not suggesting it.

```
=== RUN   TestLocalFileDataSource/0
    data_source_local_file_test.go:33: Step 1/1 error: Error running pre-apply plan: exit status 1
        
        Error: open local_file: no such file or directory
        
          on terraform_plugin_test.tf line 6, in data "local_file" "file":
           6: 				data "local_file" "file" {
        
        
--- FAIL: TestLocalFileDataSource (1.91s)
    --- FAIL: TestLocalFileDataSource/0 (1.91s)
```

Windows - All Versions:

The testing appears to be using `filepath.Join()` in the test functions, which normally would be correct, however the value is being inserted into a Terraform string so any forward slashes will be treated as escape patterns. It may work if we simply replace `\` characters with `/` or `\\` in the test functions.

```
=== RUN   TestLocalFile_Basic/0
    resource_local_file_test.go:64: Step 1/1 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
    testing_new.go:56: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
        
=== RUN   TestLocalFile_Basic/1
    resource_local_file_test.go:64: Step 1/1 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
    testing_new.go:56: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
        
=== RUN   TestLocalFile_Basic/2
    resource_local_file_test.go:64: Step 1/1 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
    testing_new.go:56: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
        
=== RUN   TestLocalFile_Basic/3
    resource_local_file_test.go:64: Step 1/1 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
    testing_new.go:56: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4:   filename = "C:\Users\RUNNER~1\AppData\Local\Temp\tf006894185\local_file"
        
        The symbol "l" is not a valid escape sequence selector.
        
--- FAIL: TestLocalFile_Basic (11.37s)
    --- FAIL: TestLocalFile_Basic/0 (2.84s)
    --- FAIL: TestLocalFile_Basic/1 (2.83s)
    --- FAIL: TestLocalFile_Basic/2 (2.89s)
    --- FAIL: TestLocalFile_Basic/3 (2.81s)

=== RUN   TestLocalFile_Permissions
    resource_local_file_test.go:155: Step 1/1 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "T" is not a valid escape sequence selector.
    testing_new.go:56: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The \U escape sequence must be followed by eight hexadecimal digits.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "R" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "A" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "L" is not a valid escape sequence selector.
        
        
        Error: Invalid escape sequence
        
          on terraform_plugin_test.tf line 4, in resource "local_file" "file":
           4: 	filename             = "C:\Users\RUNNER~1\AppData\Local\Temp\tf692227977/local_file"
        
        The symbol "T" is not a valid escape sequence selector.
        
--- FAIL: TestLocalFile_Permissions (2.75s)
```

**Comment by kmoe** [comment]
Fabulous - thanks for the investigation, @bflad. Please check the fixes are satisfactory and re-review.

**Review by bflad — APPROVED** [review]
This looks great! 🚀

**Pull request merged**